File: google/main/src/core/apps/SearchEngineApp.java
package main.src.core.apps;

import main.src.core.structures.Document;
import main.src.utils.FileHandler;
import main.src.utils.Prettifier;
import main.src.core.query.Query;
import main.src.utils.ConsoleApp;
import main.src.core.engine.InvertedIndex;
import main.src.core.engine.Engine;

import java.util.*;
import java.lang.IllegalArgumentException;


public class SearchEngineApp extends ConsoleApp{

    private static final String APP_NAME = "google";
    private final String YEAR = "2020";
    private final String VERSION = "v0.0.0";
    
    private String resourcesDirectory;
    private InvertedIndex index;

    public void intro(){
        this.sout("Welcome to " + APP_NAME + "! Copyright(c) " + YEAR + " " + VERSION);
    }

    public SearchEngineApp(String resourcesDirectory){
        super();
        this.resourcesDirectory = resourcesDirectory;
        this.index = new InvertedIndex(this.resourcesDirectory);
        this.prompt = APP_NAME + "> ";
    }

    public boolean handleCommand(String command, String arguments){
        if (command.equals("search"))    this.search(arguments);
        else if (command.equals("view")) this.view(arguments);
        else if (command.equals("help")) this.help();
        else if (command.equals("exit")) return false;
        else this.sout(String.format("\'%s\' is not recognized as an internal or external command.", command));
        return true;
    }

    private String getSearchResults(String arguments){
        HashSet<Document> results = Engine.getQueryResults(new Query(arguments), this.index);
        if(results.isEmpty())
            return "No results found!";
        return Prettifier.prettify(results);
    }

    public void search(String arguments) {
        if(arguments == ""){
            this.sout("No keywords passed!");
            return;
        }
        try {
            this.sout(this.getSearchResults(arguments));
        } catch (IllegalArgumentException e) {
            this.sout(e.getMessage());
        }
    }

    public void view(String arguments) {
        if(arguments == ""){
            this.sout("No documentId passed!");
            return;
        }
        this.sout(FileHandler.loadFile(this.resourcesDirectory + "/" + arguments));
    }

    public void help() {
        this.sout("search <terms> -- view <documentId> -- help");
    }

    private void sout(String output){
        System.out.println("\n" + output + "\n");
    }

}

File: GoogleSharpApp/SearchEngine/SearchEngineApp.cs
using System;
using System.Collections.Generic;
using System.Linq;
using GoogleSharp.Src.Core.Engine;
using GoogleSharp.Src.Core.Query;
using GoogleSharp.Src.Core.Structures;
using GoogleSharp.Src.Utils;

namespace SearchEngine {
    public class SearchEngineApp : ConsoleApp {
        private const string AppName = "google";
        private const string Year = "2020";
        private const string Version = "v1.0.1";

        private readonly string ResourcesDirectory;
        private GoogleInvertedIndex index;

        public override void Intro() {
            WriteLine("Welcome to " + AppName + "! Copyright(c) Team8 " + Year + " " + Version);
        }

        public SearchEngineApp(string resourcesDirectory) : base() {
            ResourcesDirectory = resourcesDirectory;
            var fileHandler = new FileHandler();
            index = new GoogleInvertedIndex(fileHandler.GetDocumentsFromFolder(ResourcesDirectory), fileHandler);
            prompt = AppName + "> ";
        }

        public override bool HandleCommand() {
            if (command.Equals("search")) Search(arguments);
            else if (command.Equals("view")) View(arguments);
            else if (command.Equals("help")) Help();
            else if (command.Equals("exit")) return false;
            else WriteLine($"'{command}' is not recognized as an internal or external command.");
            return true;
        }

        private string GetSearchResults(string arguments) {
            HashSet<Document> results = QueryEngine.GetQueryResults(new QueryBuilder(arguments), index);
            if (!results.Any())
                return "No results found!";
            return Prettifier<Document>.Prettify(results);
        }

        public void Search(string arguments) {
            if (arguments == "") {
                WriteLine("No keywords passed!");
                return;
            }
            try {
                WriteLine(GetSearchResults(arguments));
            } catch (ArgumentException e) {
                WriteLine(e.Message);
            }
        }

        public void View(string arguments) {
            if (arguments == "") {
                WriteLine("No documentId passed!");
                return;
            }
            WriteLine(new FileHandler().GetFileContent(ResourcesDirectory + "/" + arguments));
        }

        public override void Help() {
            WriteLine("search <terms> -- view <documentId> -- help");
        }

        private void WriteLine(string output) {
            Console.WriteLine("\n" + output + "\n");
        }
    }
}
